executor/mem_reader.go

```diff
@@ -202,7 +202,15 @@ func buildMemTableReader(ctx context.Context, us *UnionScanExec, tblReader *Tabl
 	if len(pkColIDs) == 0 {
 		pkColIDs = []int64{-1}
 	}
-	rd := rowcodec.NewByteDecoder(colInfo, pkColIDs, nil, us.ctx.GetSessionVars().Location())
+
+	defVal := func(i int) ([]byte, error) {
+		d, err := table.GetColOriginDefaultValue(us.ctx, us.columns[i])
+		if err != nil {
+			return nil, err
+		}
+		return tablecodec.EncodeValue(us.ctx.GetSessionVars().StmtCtx, nil, d)
+	}
+	rd := rowcodec.NewByteDecoder(colInfo, pkColIDs, defVal, us.ctx.GetSessionVars().Location())
```

Comment:
Good catch, also need change code below for `indexLookUp` and `indexMerge`.
https://github.com/pingcap/tidb/blob/8b7b31005148b14b92f63b9a46cac0dcbc5c7c6f/executor/mem_reader.go#L795

And could add a UT like
```
create table cache_add_column (f1 int, index k(f1));
insert into cache_add_column (f1) values (1);
alter table cache_add_column add column f2 int not null, add column f3 int default 3, add column f4 int default null;
alter table cache_add_column cache;
select * from cache_add_column use index(k) where f1 = 1;
select * from cache_add_column use index(k) where f1 = 1;
```

Reply:
done, added a new commit